### PR TITLE
Added new KTL logic to KeepDecorator

### PR DESCRIPTION
**Pull request by ryanpbrewster**
This makes it so that the PageCommander can access multilib keeps (which in
turn makes them available to the extension).

@andrewconner @leogrim

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala [hunk 1]
```diff
@@ -192,11 +204,19 @@ class KeepDecorator @Inject() (
...
-      val allKeeps = keepRepo.getByLibraryIdsAndUriIds(writeableLibs, uriIds)
-      allKeeps
+      val oldWay = keepRepo.getByLibraryIdsAndUriIds(writeableLibs, uriIds).toSet
+      val newWay = {
+        val direct = ktuRepo.getByUserIdAndUriIds(userId, uriIds).map(_.keepId)
+        val indirectViaLibraries = ktlRepo.getVisibileFirstOrderImplicitKeeps(uriIds, writeableLibs).map(_.keepId)
+        (direct ++ indirectViaLibraries) |> keepRepo.getByIds |> (_.values.toSet)
+      }
+      if (newWay.map(_.id.get) != oldWay.map(_.id.get)) {
+        log.error(s"[KTL-MATCH] getBasicKeeps($userId, $uriIds): ${newWay.map(_.id.get)} != ${oldWay.map(_.id.get)}")
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala:217]: Maybe even airbrake for now?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala [hunk 2]
```diff
@@ -192,11 +204,19 @@ class KeepDecorator @Inject() (
...
-  def getBasicKeeps(userId: Id[User], uriIds: Set[Id[NormalizedURI]]): Map[Id[NormalizedURI], Set[BasicKeep]] = {
+  def getBasicKeeps(userId: Id[User], uriIds: Set[Id[NormalizedURI]], useMultilibLogic: Boolean = false): Map[Id[NormalizedURI], Set[BasicKeep]] = {
     val allKeeps = db.readOnlyReplica { implicit session =>
       val writeableLibs = libraryMembershipRepo.getLibrariesWithWriteAccess(userId)
-      val allKeeps = keepRepo.getByLibraryIdsAndUriIds(writeableLibs, uriIds)
-      allKeeps
+      val oldWay = keepRepo.getByLibraryIdsAndUriIds(writeableLibs, uriIds).toSet
+      val newWay = {
+        val direct = ktuRepo.getByUserIdAndUriIds(userId, uriIds).map(_.keepId)
+        val indirectViaLibraries = ktlRepo.getVisibileFirstOrderImplicitKeeps(uriIds, writeableLibs).map(_.keepId)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepDecorator.scala:213]: Typo in `getVisibileFirstOrderImplicitKeeps`. 
Is there a reason not to have a single super-join or `union` query to grab all keeps by `uriId` for a given user?
I guess one advantage is that you can have the direct ones at the top of the list here.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala [hunk 3]
```diff
@@ -127,17 +127,12 @@ class KeepToLibraryRepoImpl @Inject() (
     getByUserIdAndLibraryIdHelper(userId, libraryId, excludeStateOpt).list
   }
 
-  def getVisibileFirstOrderImplicitKeeps(userId: Id[User], uriId: Id[NormalizedURI])(implicit session: RSession): Set[Id[Keep]] = {
+  def getVisibileFirstOrderImplicitKeeps(uriIds: Set[Id[NormalizedURI]], libraryIds: Set[Id[Library]])(implicit session: RSession): Set[KeepToLibrary] = {
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:130]: Why not just `getKeepsByUriIdsAndLibraryIds`?
Or `getKeepsByUrisInLibraries`?

(this method is viewer independent, so speaking of degree here doesn't make much sense, what do you think?)

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala:130]: Those are both better names. The naming is a holdover from when I was doing
a monster join on org memberships as well.

On Mon, Aug 24, 2015 at 6:34 PM, Léo Grimaldi notifications@github.com
wrote:

> In
> kifi-backend/modules/shoebox/app/com/keepit/model/KeepToLibraryRepo.scala
> https://github.com/kifi/keepit/pull/16955#discussion_r37824778:
> 
> > @@ -127,17 +127,12 @@ class KeepToLibraryRepoImpl @Inject() (
> >      getByUserIdAndLibraryIdHelper(userId, libraryId, excludeStateOpt).list
> >    }
> > -  def getVisibileFirstOrderImplicitKeeps(userId: Id[User], uriId: Id[NormalizedURI])(implicit session: RSession): Set[Id[Keep]] = {
> > -  def getVisibileFirstOrderImplicitKeeps(uriIds: Set[Id[NormalizedURI]], libraryIds: Set[Id[Library]])(implicit session: RSession): Set[KeepToLibrary] = {
> 
> Why not just getKeepsByUriIdsAndLibraryIds?
> Or getKeepsByUrisInLibraries?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/kifi/keepit/pull/16955/files#r37824778.

## 

Ryan Brewster, Kifi http://kifi.com/
Software Engineer

**Pull request merged**